annnufan/assignment4/islands_counter.cpp

```diff
@@ -0,0 +1,59 @@
+#include <vector>
+#include <utility>
+
+class islands_counter {
+	std::vector<std::vector<bool>> island_map;
+	std::vector<std::pair<int, int>> moves;
+	int n, m;
+	int islands;
+	bool is_count;
+
+	void dfs(int i, int j) {
+		island_map[i][j] = false;
+
+		for (const auto& move : moves) {
+			int i1 = move.first + i;
+			int j1 = move.second + j;
+			if (0 <= i1 && i1 < n && 0 <= j1 && j1 < m && island_map[i1][j1]) {
+				dfs(i1, j1);
+			}
+		}
+	}
+
+public:
+	islands_counter(std::vector<std::vector<bool>> i_m) : island_map(i_m) {
+		n = island_map.size();
+		m = (n != 0) ? island_map[0].size() : 0;
+		islands = 0;
+		is_count = false;
+		for (int i = -1; i <= 1; i++) {
+			for (int j = -1; j <= 1; j++) {
+				if ((i == 0) ^ (j == 0)) {
+					moves.push_back({i, j});
+				}
+			}
+		}
+	}
+
+	int count() {
+		if (is_count) {
+			return islands;
+		}
+		is_count = true;
+		for (int i = 0; i < n; i++) {
+			for (int j = 0; j < m; j++) {
+				if (island_map[i][j]) {
+					dfs(i, j);
+					islands++;
+				}
+			}
+		}
+		return islands;
+	}
+
+};
+
+int count_islands_on_map(std::vector<std::vector<bool>> island_map) {

```

Comment:
Tip:
Encapsulating this in a function is perfectly fine (it actually IS a simple function - get the island map, return the number of island). For that matter, even the existence of this class is an implementation detail! Not bad :)

Other options (just for your information):

You can implement this as a one liner, if you choose:
```C++
return islands_counter(island_map).count();
```
Or you can remove call the one liner instead of this function in the tests:
```C++
EXPECT_EQ(islands_counter(island_map).count(), 4);
```
saves this function and still readable IMO.

All of these options, including what you did, is OK, choose what you think is most readable.